README.md
# list_timeline

Plugin to create a simple tracking widget for iOS and Android, Windows, and Web
<br><br>
<img src="https://raw.githubusercontent.com/griajobag/list_timeline/main/timeline.png"/> 

## Usage

To use this plugin, add ```list_timeline``` as
a [dependency in your pubspec.yaml](https://flutter.io/platform-plugins/).

### Example

```dart
import 'package:example/data_model.dart';
import 'package:flutter/material.dart';
import 'package:intl/intl.dart';
import 'package:list_timeline/custom_list_tracking.dart';

void main() => runApp(App());

class App extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'List Timeline',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({
    Key? key,
  }) : super(key: key);

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  List<DataModel> listExample = [];

  @override
  void initState() {
    listExample.add(DataModel(
        title: "Approved",
        desc: "This task was approved by your manager",
        dateTime: DateTime(2022, 08, 10)));

    listExample.add(DataModel(
        title: "Warning",
        desc: "This task was got yellow notice by your manager",
        dateTime: DateTime(2022, 08, 12)));

    listExample.add(DataModel(
        title: "Rejected",
        desc: "This task was rejected by your manager",
        dateTime: DateTime(2022, 08, 23)));
    super.initState();
  }

  String _dateFormat(DateTime date) {
    return DateFormat("dd/MM/yyyy").format(date);
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text("List Timeline Usage"),
      ),
      //This below codes are the simple example how to use the package
      // if you wanna custom widget, just find the following attribute :
      // 1. customLeftWidget: (e)=>Container(),
      // 2. customTitleWidget: (e) => Container(),
      // 3. customDescWidget: (e) => Container(),
      
      body: Container(
          margin: const EdgeInsets.all(10),
          child: CustomListTracking<DataModel>(
            listItem: listExample,
            valueTextOfTitle: (e) => e.title!,
            valueTextOfDesc: (e) => e.desc!,
            colorCircleTimeline: (e) =>
            e.title == "Warning" ? Colors.yellow : e.title == "Rejected"
                ? Colors.red
                : Colors.blue,
            showLeftWidget: true,
            valueOfLeftSource: (e) => _dateFormat(e.dateTime!),
          )),
    );
  }
}
```

CHANGELOG.md
## 0.0.1 - 2022 - 08 - 24
* Initial publish package
## 0.0.2 - 2022 - 08 - 24
* Add image to readme
## 0.0.3 - 2022 - 08 - 24
* Add image to readme fixed

test/list_timeline_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:list_timeline/custom_list_tracking.dart';

void main() {

}


lib/image_component.dart
import 'dart:io';

import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';

class ImageComponent extends StatefulWidget {
  final String? imageSource;
  final bool isNetworkImage;
  final bool isCircleImage;
  final double height;
  final double width;
  final double textSize;
  final dynamic isImageFromFile;
  final EdgeInsets edgeInsets;

  const ImageComponent(
      {Key? key,
        required this.imageSource,
        this.isNetworkImage = true,
        this.height = 100,
        this.width = 100,
        this.textSize = 12,
        this.edgeInsets = const EdgeInsets.all(10),
        this.isCircleImage = false,
        this.isImageFromFile})
      : super(key: key);

  @override
  State<ImageComponent> createState() => _ImageComponentState();
}

class _ImageComponentState extends State<ImageComponent> {
  @override
  Widget build(BuildContext context) {
    return Container(
      margin: widget.edgeInsets,
      child: widget.isNetworkImage
          ? ClipRRect(
        borderRadius: widget.isCircleImage
            ? BorderRadius.circular(100)
            : BorderRadius.circular(0),
        child: Image.network(
          widget.imageSource!,
          height: widget.height,
          width: widget.width,
          fit: BoxFit.cover,
          errorBuilder: (context, error, stackTrace) =>
              Text(error.toString(), style: TextStyle(fontSize: widget.textSize),),
        ),
      )
          : widget.isImageFromFile != null
          ? ClipRRect(
          borderRadius: widget.isCircleImage
              ? BorderRadius.circular(100)
              : BorderRadius.circular(0),
          child: kIsWeb
              ? Image.memory(
            widget.isImageFromFile.bytes,
            width: widget.width,
            height: widget.height,
            fit: BoxFit.cover,
          )
              : Image.file(
            File(widget.isImageFromFile!.path),
            width: widget.width,
            height: widget.height,
            fit: BoxFit.cover,
          ))
          : ClipRRect(
          borderRadius: widget.isCircleImage
              ? BorderRadius.circular(100)
              : BorderRadius.circular(0),
          child: Image(
            image: AssetImage(widget.imageSource!),
            height: widget.height,
            width: widget.width,
            fit: BoxFit.cover,
          )),
    );
  }
}


lib/custom_list_tracking.dart
library list_timeline;

import 'package:flutter/material.dart';
import 'package:list_timeline/image_component.dart';

class CustomListTracking<T> extends StatelessWidget {
  final List<T>? listItem;
  final String Function(T)? valueTextOfTitle;
  final String Function(T)? valueTextOfDesc;
  final String Function(T)? valueOfLeftSource;
  final bool showLeftWidget;
  final Color Function(T)? colorCircleTimeline;
  final bool isPrimary;
  final double textSize;
  final double textSizeOfDateLeftTimeline;
  final Widget Function(T)? customTitleWidget;
  final Widget Function(T)? customDescWidget;
  final Widget Function(T)? customLeftWidget;

  const CustomListTracking({
    Key? key,
    required this.listItem,
    this.isPrimary = false,
    this.valueTextOfTitle,
    this.valueTextOfDesc,
    this.valueOfLeftSource,
    this.colorCircleTimeline,
    this.showLeftWidget = false,
    this.customTitleWidget,
    this.customDescWidget,
    this.customLeftWidget,
    this.textSize = 12,
    this.textSizeOfDateLeftTimeline = 12,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return ListView.builder(
        shrinkWrap: !isPrimary,
        primary: isPrimary,
        itemCount: listItem!.length,
        itemBuilder: (context, index) => _itemTimeline(index));
  }

  Widget _itemTimeline(
    int currentIndex,
  ) {
    return IntrinsicHeight(
      child: Row(
        mainAxisAlignment: MainAxisAlignment.start,
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          Visibility(
              visible: showLeftWidget,
              child: Padding(
                padding: const EdgeInsets.only(top: 3, left: 5, right: 5),
                child: customLeftWidget == null
                    ? Text(
                        (valueOfLeftSource == null)
                            ? ""
                            : valueOfLeftSource!(listItem![currentIndex]),
                        style: TextStyle(fontSize: textSizeOfDateLeftTimeline),
                      )
                    : customLeftWidget!(listItem![currentIndex]),
              )),
          Column(
            mainAxisAlignment: MainAxisAlignment.start,
            crossAxisAlignment: CrossAxisAlignment.center,
            children: [
              Container(
                width: 22,
                height: 22,
                margin:
                    const EdgeInsets.symmetric(vertical: 0.0, horizontal: 0.0),
                decoration: BoxDecoration(
                  border: Border.all(
                      color: colorCircleTimeline == null
                          ? Colors.blue
                          : colorCircleTimeline!(listItem![currentIndex]),
                      width: 2.5,
                      style: BorderStyle.solid),
                  shape: BoxShape.circle,
                ),
                child: Container(
                  width: double.infinity,
                  height: double.infinity,
                  margin: const EdgeInsets.all(2),
                  decoration: BoxDecoration(
                    color: colorCircleTimeline == null
                        ? Colors.blue
                        : colorCircleTimeline!(listItem![currentIndex]),
                    shape: BoxShape.circle,
                  ),
                ),
              ),
              (currentIndex + 1 != listItem!.length)
                  ? Expanded(
                      child: Container(
                        height: double.infinity,
                        width: 2.5,
                        color: colorCircleTimeline == null
                            ? Colors.blue
                            : colorCircleTimeline!(listItem![currentIndex]),
                      ),
                    )
                  : Container()
            ],
          ),
          Expanded(
            child: Padding(
              padding: const EdgeInsets.only(left: 12),
              child: Column(
                mainAxisAlignment: MainAxisAlignment.start,
                crossAxisAlignment: CrossAxisAlignment.start,
                children: [
                  const SizedBox(height: 2),
                  customTitleWidget == null
                      ? Text(
                          valueTextOfTitle == null
                              ? ""
                              : valueTextOfTitle!(listItem![currentIndex]),
                          style: const TextStyle(fontWeight: FontWeight.bold),
                        )
                      : customTitleWidget!(listItem![currentIndex]),
                  customDescWidget == null
                      ? Padding(
                          padding: const EdgeInsets.only(top: 4, bottom: 12),
                          child: Text(valueTextOfDesc == null
                              ? ""
                              : valueTextOfDesc!(listItem![currentIndex])),
                        )
                      : customDescWidget!(listItem![currentIndex]),
                ],
              ),
            ),
          ),
        ],
      ),
    );
  }
}


example/lib/main.dart
import 'package:example/data_model.dart';
import 'package:flutter/material.dart';
import 'package:intl/intl.dart';
import 'package:list_timeline/custom_list_tracking.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'List Timeline',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({
    Key? key,
  }) : super(key: key);

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  List<DataModel> listExample = [];

  @override
  void initState() {
    listExample.add(DataModel(
        title: "Approved",
        desc: "This task was approved by your manager",
        dateTime: DateTime(2022, 08, 10)));

    listExample.add(DataModel(
        title: "Warning",
        desc: "This task was got yellow notice by your manager",
        dateTime: DateTime(2022, 08, 12)));

    listExample.add(DataModel(
        title: "Rejected",
        desc: "This task was rejected by your manager",
        dateTime: DateTime(2022, 08, 23)));
    super.initState();
  }
  
  String _dateFormat(DateTime date){
    return DateFormat("dd/MM/yyyy").format(date);
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text("List Timeline Usage"),
      ),
      body: Container(
          margin: const EdgeInsets.all(10),
          child: CustomListTracking<DataModel>(
            listItem: listExample,
            valueTextOfTitle: (e) => e.title!,
            valueTextOfDesc: (e) => e.desc!,
            colorCircleTimeline:(e)=> e.title=="Warning"?Colors.yellow: e.title=="Rejected"?Colors.red:Colors.blue   ,
            showLeftWidget: true,
            customLeftWidget: (e)=>Column(
              children: [
                Text(_dateFormat(e.dateTime!)),
                Text(e.title!)
              ],
            ),
            valueOfLeftSource: (e) => _dateFormat(e.dateTime!),
          )),
    );
  }
}
